File: NameBST.cpp
#pragma once
#include "NameBST.h"
#include <iostream>
#include <string>
#include <fstream>

using namespace std;

NameBST::NameBST() { root = nullptr; }
NameBST::~NameBST() { 
	while (root != nullptr)
		remove(root);
}

NameBSTNode* NameBST::getRoot() { return root; }	//Get the address of root.

void NameBST::insert(NameBSTNode* node) {
/*//////////////////////////////////////////////////////////////////////////
                function name : insert
        Input : NameBSTNode* - the node wants to insert into the tree.
        Output : None.
        Purpose : Insert a node into the tree.
//////////////////////////////////////////////////////////////////////////*/
	if (root == nullptr)
		root = new NameBSTNode(node->getName(), node->getAge(), node->getSubDate(), node->getSubEndDate());
	//If there's no node in the tree, make a new node and set it as a root.
	else {
	//If not, executes following codes.
		NameBSTNode* prev;
		NameBSTNode* it = root;
		int result = 0;

		while (it!= nullptr) {
			prev = it;
			//Save the address of the parent node of the node you want to move.
			if ((result = node->getName().compare(it->getName())) == 0) {
			/*	If the name of the visited node and the name of the node passed
				as an argument are the same, execute the code below		*/
				if(it->getAge() == node->getAge()) {
					if (it->getSubDate().compare(node->getSubDate()) == 0) {
						if (it->getSubEndDate().compare(node->getSubEndDate()) == 0)
							return;
				//If all information on the visited node and the received node are the same, ends the function.
						else
							it = it->getRight();
					}
					else
						it = it->getRight();
				}
				else
					it = it->getRight();
				//If not, move "it" to the right child.
			}
			else if (result < 0)
				it = it->getLeft();
			/*	If the name of the visited node is faster in lexicographic order,
				"it" moves to the left child of that node.	*/
			else
				it = it->getRight();
			/*	If the delivered one is slower, moves to the right child and save the parent node's address.	*/
		}
		it = new NameBSTNode(node->getName(), node->getAge(), node->getSubDate(), node->getSubEndDate());
		/*	Creates a node based on the input information.	*/
		it->setParent(prev);
		/*	Set the parent of new node.	*/
		if (result < 0)
			prev->setLeft(it);
		/*	If the name of the parent node is faster in lexicographic order,
			set the new node as a parent's left child.*/
		else
			prev->setRight(it);
		//If not, set it as a right child.
	}
	return;
}
void NameBST::print(NameBSTNode* node) {
/*//////////////////////////////////////////////////////////////////////////
                function name : print
        Input : NameBSTNode* - the node wants to print.
        Output : None.
        Purpose : Prints all nodes in the tree recursively.
//////////////////////////////////////////////////////////////////////////*/
	if (node != nullptr) {
	//Executes following code if the delivered node is not a null value.
		print(node->getLeft());
		//Call the other function for its left child.
		ofstream flog;
		flog.open("log.txt", ios::app);
		flog << node->getName() << "/" << node->getAge() << "/" << node->getSubDate() << "/" << node->getSubEndDate() << endl;
		flog.close();
		//If the node's left subtree's information is printed, print myself. 
		print(node->getRight());
		//Call the other function for its right child.
	}
	//Finally, all nodes in the tree are output in inorder.
	return;
}
void NameBST::remove(NameBSTNode* node) {
/*//////////////////////////////////////////////////////////////////////////
                function name : remove
        Input : NameBSTNode* - the node wants to remove.
        Output : None.
        Purpose : Delete the node and restruct the tree.
//////////////////////////////////////////////////////////////////////////*/
	if (node == nullptr)
		return;	//If delivered nullptr, terminate the function.
	NameBSTNode* it = nullptr;
	if (node->getLeft() != nullptr) {
		if (node->getRight() != nullptr) {
			it = node->getRight();
			while (it->getLeft() != nullptr)
				it = it->getLeft();
		}
		/*	If degree of node that want to delete is 2,
			searches for the leftmost node in the right subtree of the node.	*/
		else
			it = node->getLeft();
		//If the node only has left chid, set the replacement node as the left child.
	}
	else {
		if (node->getRight() != nullptr)
			it = node->getRight();
		//If only has right, set it as the right child.
	}

	if (it == nullptr) {
		//If the node is leaf node, run the following codes.
		if (node != root) {
		//If the tree has more than two nodes, executes the following codes.
			if (node->getParent()->getLeft() == node)
				node->getParent()->setLeft(nullptr);
			else
				node->getParent()->setRight(nullptr);
		//	Determine which node the node is a child of.
			delete node;
			return;
		//	Terminate the function after delete the node.
		}
		else {
			delete node;
			root = nullptr;
			return;
		/*	If the tree has only one node, set root as nullptr value
		and terminate the function after delete the node.	*/
		}
	}
	else {
	//If the degree is over 0, executes following codes.
		if (it->getParent()->getLeft() == it)
			it->getParent()->setLeft(nullptr);
		else if (it->getParent()->getRight() == it)
			it->getParent()->setRight(nullptr);
		/*	Determine which child of the parent node the replacement node is,
			and change the corresponding node to a null value.	*/
		if (node != root) {
			if (node->getParent()->getLeft() == node)
				node->getParent()->setLeft(it);
			else if (node->getParent()->getRight() == node)
				node->getParent()->setRight(it);
			it->setParent(node->getParent());
		/*	If the node wants to delete is not a root, determine which child of the parent node it is
			and replace that node with a replacement node "it".
			And set the "it"'s parent to the parent of the node wants to delete.	*/
		}
		else {
			if (it->getParent() == node->getLeft()) {
				if (node->getLeft()->getLeft() == it)
					node->getLeft()->setLeft(nullptr);
				else if (node->getLeft()->getRight() == it)
					node->getLeft()->setRight(nullptr);
			}
			/* If the node is root, and if the grand parent of “it” is "node", determine the direction
				of "it", and change the node in the corresponding position to a null value.*/
			else if (it->getParent() == node->getRight()) {
				if (node->getRight()->getLeft() == it)
					node->getRight()->setLeft(nullptr);
				else if (node->getRight()->getRight() == it)
					node->getRight()->setRight(nullptr);
			/*	if "it"'s parent is a node's right child,
				determine and change the corresponding position in same way.	*/
			}
			it->setParent(nullptr);
			root = it;
			//Set the "it" as a root.
		}
	}
	if (node->getLeft() != nullptr && node->getLeft() != it) {
		it->setLeft(node->getLeft());
		node->getLeft()->setParent(it);
	//	Change the left child of the node want to delete to the left child of "it".
	}
	if (node->getRight() != nullptr && node->getRight() != it) {
		it->setRight(node->getRight());
		node->getRight()->setParent(it);
	}
	//	Change the "it"'s right child in the same way.
	delete node;
	node = nullptr;
	return;
	//	Delete the node and exit the function.
}
NameBSTNode* NameBST::search(const string name) {
/*//////////////////////////////////////////////////////////////////////////
                function name : search
        Input : const string	- Name wants to find.
        Output : NameBSTNode* - address : Node corresponding to the entered name.
				 			  - nullptr 	: No node corresponding in the tree.
        Purpose : Find the node named with "name" in the tree.
//////////////////////////////////////////////////////////////////////////*/
	NameBSTNode* it = root;
	//	Set the compare position of the tree as a root.
	int result = 0;
	while (it!= nullptr) {
        result = name.compare(it->getName());
        //Compare the two names that are one for entered, and one for already existed in the tree.
        if (result == 0) 
            break;
        //  If two names are same, break out of the loop.
        else if (result < 0)
            it = it->getLeft();
        /*  If the two names are different, and name delivered is alphabetically earlier,
            than move "it" to the left child.     */
        else
            it = it->getRight();
        //  If not, move "it" to the right child.
    }
	return it;
	//Return the current address of it.
}

File: javaVersion/src/lib/NameBST.java
package lib;

import java.io.BufferedWriter;
import java.io.FileNotFoundException;
import java.io.FileWriter;
import java.io.IOException;

public class NameBST {
    private BSTNode root;

    public NameBST() {root = null;}
    public BSTNode getRoot() {return root;}
    public void insert(BSTNode node) {
        if (root == null) {
            root = new BSTNode(node.getName(), node.getAge(),
                    node.getSubDate(), node.getSubEndDate());
        }
        else {
            BSTNode prev = null, it = root;
            String it_name, it_sub_date, it_sub_end,
                    node_name = node.getName(), node_sub_date = node.getSubDate(),
                    node_sub_end = node.getSubEndDate();
            int it_age, node_age = node.getAge(), result = 0;

            while (it != null) {
                prev = it;
                it_name = it.getName();
                it_sub_date = it.getSubDate();
                it_sub_end = it.getSubEndDate();
                it_age = it.getAge();

                result = it_name.compareTo(node.getName());
                if (result != 0) {
                    if (result > 0) it = it.getLeft();
                    else it = it.getRight();
                    continue;
                }
                else {
                    if (it_age == node_age);
                    else {
                        if (it_age < node_age) it = it.getRight();
                        else it = it.getLeft();
                        continue;
                    }

                    result = it_sub_date.compareTo(node_sub_date);
                    if (result == 0);
                    else {
                        if (result < 0) it = it.getRight();
                        else it = it.getLeft();
                        continue;
                    }

                    result = it_sub_end.compareTo(node_sub_end);
                    if (result == 0) return;
                    else {
                        if (result < 0) it = it.getRight();
                        else it = it.getLeft();
                        continue;
                    }
                }
            }
            it = new BSTNode(node_name, node_age, node_sub_date, node_sub_end);
            it.setParent(prev);
            if (result > 0) prev.setLeft(it);
            else prev.setRight(it);
        }
    }
    public BSTNode search(String name) {
        BSTNode it = root;
        while (it != null) {
            if (it.getName().compareTo(name) == 0) break;
            else if (it.getName().compareTo(name) < 0) it = it.getRight();
            else it = it.getLeft();
        }
        return it;
    }
    public void print(BSTNode node) {
        if (node.getLeft() != null)
            print(node.getLeft());
        String contents = node.getName() + "/" + node.getAge() + "/" + node.getSubDate() + "/" + node.getSubEndDate() + "\n";
        try {
            BufferedWriter br = new BufferedWriter(new FileWriter("./log.txt", true));
            br.write(contents);
            br.close();
        } catch (FileNotFoundException e) {
            System.out.println("File not found");
            e.printStackTrace();
        } catch (IOException e) {
            System.out.println("Error writing to log file");
            e.printStackTrace();
        }
        if (node.getRight() != null)
            print(node.getRight());
    }
    public void remove(BSTNode node) {
        if (node == null) return;

        BSTNode it = null, temp, node_left_child = node.getLeft(), node_right_child = node.getRight();
        int degree = 0;
        if (node_left_child != null) degree++;
        if (node_right_child != null) degree++;

        if (degree == 2) {
            it = node.getRight();
            degree = 0;
            while (it.getLeft() != null) {
                it = it.getLeft();
                degree++;
            }
            // 삭제할 노드의 오른쪽 서브트리에서 키값이 가장 작은 자식으로 이동.
            temp = new BSTNode(it.getName(), it.getAge(), it.getSubDate(), it.getSubEndDate());
            temp.setLeft(node.getLeft());
            node.getLeft().setParent(temp);
            // 삭제할 노드의 왼쪽 자식과 대체할 노드를 연결.
            if (degree > 1) {
                temp.setRight(node.getRight());
                node.getRight().setParent(temp);
            }
            // 오른쪽 서브트리의 height 가 2 이상일 때 대체할 노드와 서브트리의 루트를 연결.
            if (it.getRight() != null) {
                it.getParent().setLeft(it.getRight());
                it.getRight().setParent(it);
            }
            // 대체할 노드에 오른쪽 자식이 있다면 대체 노드의 원래 부모와 자식을 연결.
            it = node.getParent();
            if (it == null) root = temp;
            else {
                if (it.getLeft() == node) it.setLeft(temp);
                else it.setRight(temp);
                temp.setParent(it);
            }
            // 삭제할 노드가 루트라면 대체할 노드를 루트로 설정, 아니라면 삭제할 노드의 부모와 자신을 연결.
        }
        else if (degree == 1) {
            // degree = 1
            if (node_left_child != null) it = node.getLeft();
            else it = node.getRight();

            it.setParent(node.getParent());
            if (node == root) root = it;
            else if (node.getParent().getLeft() == node) node.getParent().setLeft(it);
            else node.getParent().setRight(it);

        }
        else { // degree = 0
            it = node.getParent();
            if (it != null) {
                if (node == it.getLeft()) it.setLeft(null);
                else it.setRight(null);
            }
            else root = null;
        }
        node = null;
        return;
    }

}
